### ISSUE TITLE
Timeline not reporting compute times for some GPU ops

### ISSUE THREAD

**Issue by xilenteyex** [tensorflow/tensorflow#26340]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.1 LTS
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.12.0
- Python version: Python 2.7.15rc1
- Bazel version (if compiling from source): 0.21
- GCC/Compiler version (if compiling from source): (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
- CUDA/cuDNN version: 10.0/7.4.2
- GPU model and memory: NVIDIA Corporation GP104GL [Quadro P5000] (2 GPUs)

**Describe the current behavior**
I am running the sample [RNNLM code](https://github.com/tensorflow/models/tree/master/tutorials/rnn/ptb). I just added timeline logging code in the given tutorial for training ([Modified code here](https://github.com/xilenteyex/RNNLM_ptb_timeline)) to analyze the sequence of low level operations being executed as well as the time each operation is taking. There are a lot of operations which are placed on GPU and they have just kernel queue/launch times recorded in timeline in `/job:localhost/replica:0/task:0/device:GPU:.* Compute` streams. I am unable to find actual time to execute those operations in `/device:GPU:0/.*Compute` streams. To identify these operations, I used [this script](https://github.com/xilenteyex/RNNLM_ptb_timeline/blob/master/get_unexec_ops.py). Is this a bug, or am I missing something in my understanding of Timeline? [List of operations with kernel queue times reported, but no execution times can be found here.](https://github.com/xilenteyex/RNNLM_ptb_timeline/blob/master/un_exec_tmp.txt)
I want to do this because I am trying to do model parallelism for this RNNLM (Reccurrent Neural Network for Language Modeling) code and I want to place ops after running a heuristic graph partitioning algorithm which takes as input the estimated compute time as well as dependencies for each operation.

**Describe the expected behavior**
For every kernel that is queued on a GPU I expect it to be executed at some point and logged in the Timeline

**Code to reproduce the issue**
[tutorial code for RNNLM using ptb_data_set can be found here](https://github.com/tensorflow/models/tree/master/tutorials/rnn/ptb)
[modified tutorial code with Timline logging added for training loop can be found here](https://github.com/xilenteyex/RNNLM_ptb_timeline)

execute the following commands to get the list of operations which have kernel queue times in timeline and execution times are not logged. A file named `new_unexecute_ops.txt` will be created and each line contains name of one of these operations.
 
```
git clone https://github.com/xilenteyex/RNNLM_ptb_timeline
python ptb_word_lm_timeline.py --data_path=simple-examples/data/ --model=medium
python get_unexec_ops.py logs/timeline_logs/timeline_rnnlm_med_0_100.ctf.json new_unexecute_ops.txt
```

**Other info / logs**
[Sample timeline file can be found here](https://github.com/xilenteyex/RNNLM_ptb_timeline/tree/master/logs/timeline_logs)
[List of ops with missing execution times in the timeline file can be found here](https://github.com/xilenteyex/RNNLM_ptb_timeline/blob/master/un_exec_tmp.txt)

**Comment by xilenteyex** [comment]
I am trying to write a custom graph partitioner and am stuck on this issue for a very long time. If someone can help me figure this out, it will be great. Thanks!

**Comment by ebrevdo** [comment]
+Paul Barham <pbar@google.com>

**Comment by xilenteyex** [comment]
@prb12 , can you please have a look at this ?
Thanks!

**Comment by prb12** [comment]
It's a *long* time since I worked on any of this code... but from a quick scan of the 'list of ops with missing exection times' it looks like the vast majority are what I would call 'metadata-only' ops ... i.e. ops that do not need to launch a CUDA kernel to compute their results.   Examples include things like `Shape`, `Rank`, `BroadcastGradientArgs`, which all deal with the *shapes* of tensors and return integer results in host memory.  Other ops in this list are `Const`s ... which, even if they are float tensors are already in GPU memory and all the op needs to do is return a pointer.

**Comment by xilenteyex** [comment]
Thanks for the Explanation!

**Comment by xilenteyex** [comment]
@prb12 , for a Shape type op fro example, will it copy data to GPU or it will just keep the tensor in host memory and return pointer to the reshaped tensor ?

**Comment by prb12** [comment]
A `Shape` op does not need to look at any of the values in a tensor.  The metadata lives in the hosts memory, no matter what TF device the op runs on.  It returns a tiny int32 (or int64?) tensor whose values are the dimension sizes  - these always live in host memory, even on the GPU device (one of TF's strange foibles - but it exists precisely because of things like this since much TF gradient code, does a lot of shape 'computation' and it would be dreadful to launch GPU ops to do all of this.  Also conditionals and control flow decisions need to be done on the host and mainly use int32/bool tensors).  You can look at the signatures of ops to configm this: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/shape_ops.cc#L25-L29

```
  REGISTER_KERNEL_BUILDER(Name("Shape")                          \
                              .Device(DEVICE_GPU)                \
                              .HostMemory("output")              \
                              .TypeConstraint<int32>("out_type") \
                              .TypeConstraint<type>("T"),        \
                          ShapeOp<int32>);                       \
```

Since you mentioned "reshaped tensor" ... perhaps you mean the `Reshape` op? ... in which case, the following happens.  Assuming a float tensor input on the GPU device, the Tensor metadata object is in the host's DRAM.   This contains a pointer to a buffer in GPU memory where the tensor's data lives.  `Reshape` will make a new Tensor object whose metadata contains the new shape/dimensions.  The buffer pointer can just point at the same GPU memory since no arguments to `Reshape` ever require the data to be shuffled.  (this is true of TF, pytorch, numpy etc)

Now, `Transpose` would be a different story...
